### ISSUE TITLE
IOError: Connection reset by peer

### ISSUE THREAD

**Issue by calve** [logstash-plugins/logstash-output-s3#15]
Hello.
On my configuration (Ubuntu 14.04, Logstash 1.4), ``logstash-output-s3`` randomly crash with no obvious reason.
This bug is really annoying as it crash logstash too, and thus, logs sent by hosts are lost by logstash.

Here is the stacktrace I have after the plugin crash. Please, let me know if you need more informations

```
IOError: Connection reset by peer
                       syswrite at org/jruby/ext/openssl/SSLSocket.java:667
                       do_write at file:/opt/logstash/vendor/jar/jruby-complete-1.7.11.jar!/META-INF/jruby.home/lib/ruby/shared/jopenssl19/openssl/buffering.rb:318
                          write at file:/opt/logstash/vendor/jar/jruby-complete-1.7.11.jar!/META-INF/jruby.home/lib/ruby/shared/jopenssl19/openssl/buffering.rb:336
                         write0 at file:/opt/logstash/vendor/jar/jruby-complete-1.7.11.jar!/META-INF/jruby.home/lib/ruby/1.9/net/protocol.rb:199
                          write at file:/opt/logstash/vendor/jar/jruby-complete-1.7.11.jar!/META-INF/jruby.home/lib/ruby/1.9/net/protocol.rb:173
                        writing at file:/opt/logstash/vendor/jar/jruby-complete-1.7.11.jar!/META-INF/jruby.home/lib/ruby/1.9/net/protocol.rb:190
                          write at file:/opt/logstash/vendor/jar/jruby-complete-1.7.11.jar!/META-INF/jruby.home/lib/ruby/1.9/net/protocol.rb:172
  send_request_with_body_stream at file:/opt/logstash/vendor/jar/jruby-complete-1.7.11.jar!/META-INF/jruby.home/lib/ruby/1.9/net/http.rb:1963
                           exec at file:/opt/logstash/vendor/jar/jruby-complete-1.7.11.jar!/META-INF/jruby.home/lib/ruby/1.9/net/http.rb:1929
              transport_request at file:/opt/logstash/vendor/jar/jruby-complete-1.7.11.jar!/META-INF/jruby.home/lib/ruby/1.9/net/http.rb:1325
                          catch at org/jruby/RubyKernel.java:1284
              transport_request at file:/opt/logstash/vendor/jar/jruby-complete-1.7.11.jar!/META-INF/jruby.home/lib/ruby/1.9/net/http.rb:1324
                        request at file:/opt/logstash/vendor/jar/jruby-complete-1.7.11.jar!/META-INF/jruby.home/lib/ruby/1.9/net/http.rb:1301
                        request at /opt/logstash/vendor/bundle/jruby/1.9/gems/aws-sdk-1.35.0/lib/aws/core/http/connection_pool.rb:330
                         handle at /opt/logstash/vendor/bundle/jruby/1.9/gems/aws-sdk-1.35.0/lib/aws/core/http/net_http_handler.rb:61
                    session_for at /opt/logstash/vendor/bundle/jruby/1.9/gems/aws-sdk-1.35.0/lib/aws/core/http/connection_pool.rb:129
                         handle at /opt/logstash/vendor/bundle/jruby/1.9/gems/aws-sdk-1.35.0/lib/aws/core/http/net_http_handler.rb:55
              make_sync_request at /opt/logstash/vendor/bundle/jruby/1.9/gems/aws-sdk-1.35.0/lib/aws/core/client.rb:252
            retry_server_errors at /opt/logstash/vendor/bundle/jruby/1.9/gems/aws-sdk-1.35.0/lib/aws/core/client.rb:288
              make_sync_request at /opt/logstash/vendor/bundle/jruby/1.9/gems/aws-sdk-1.35.0/lib/aws/core/client.rb:248
                 client_request at /opt/logstash/vendor/bundle/jruby/1.9/gems/aws-sdk-1.35.0/lib/aws/core/client.rb:508
             log_client_request at /opt/logstash/vendor/bundle/jruby/1.9/gems/aws-sdk-1.35.0/lib/aws/core/client.rb:390
                 client_request at /opt/logstash/vendor/bundle/jruby/1.9/gems/aws-sdk-1.35.0/lib/aws/core/client.rb:476
                return_or_raise at /opt/logstash/vendor/bundle/jruby/1.9/gems/aws-sdk-1.35.0/lib/aws/core/client.rb:372
                 client_request at /opt/logstash/vendor/bundle/jruby/1.9/gems/aws-sdk-1.35.0/lib/aws/core/client.rb:475
                     put_object at (eval):3
          write_with_put_object at /opt/logstash/vendor/bundle/jruby/1.9/gems/aws-sdk-1.35.0/lib/aws/s3/s3_object.rb:1751
                          write at /opt/logstash/vendor/bundle/jruby/1.9/gems/aws-sdk-1.35.0/lib/aws/s3/s3_object.rb:607
                write_on_bucket at /opt/logstash/lib/logstash/outputs/s3.rb:197
                         upFile at /opt/logstash/lib/logstash/outputs/s3.rb:223
                           each at org/jruby/RubyArray.java:1613
                         upFile at /opt/logstash/lib/logstash/outputs/s3.rb:215
                       register at /opt/logstash/lib/logstash/outputs/s3.rb:285
                     time_alert at /opt/logstash/lib/logstash/outputs/s3.rb:174
                           loop at org/jruby/RubyKernel.java:1521
                     time_alert at /opt/logstash/lib/logstash/outputs/s3.rb:172
```

**Comment by arabold** [comment]
I'm seeing a similar problem with Logstash 1.5.0 as well:
```
{:timestamp=>"2015-05-27T17:41:27.455000+0000", :message=>"Failed parsing date from field", :field=>"log_timestamp", :value=>"url", :exception=>java.lang.IllegalArgumentException: Invalid format: "url", :level=>:warn}
IOError: closed stream
                size at org/jruby/RubyFile.java:1124
  rotate_events_log? at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-s3-0.1.7/lib/logstash/outputs/s3.rb:287
        handle_event at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-s3-0.1.7/lib/logstash/outputs/s3.rb:327
            register at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-s3-0.1.7/lib/logstash/outputs/s3.rb:208
                call at org/jruby/RubyProc.java:271
              encode at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-codec-json_lines-0.1.8/lib/logstash/codecs/json_lines.rb:49
             receive at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-s3-0.1.7/lib/logstash/outputs/s3.rb:282
              handle at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.0-java/lib/logstash/outputs/base.rb:88
         output_func at (eval):819
        outputworker at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.0-java/lib/logstash/pipeline.rb:244
       start_outputs at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.0-java/lib/logstash/pipeline.rb:166
```

It takes whole Logstash server down with it and thus renders the plugin unusable to us. The network continues to work normally otherwise - but even if it wouldn't, I would expect the plugin to handle this gracefully and e.g. stop/pause uploading logs to S3 instead of crashing.

**Comment by calve** [comment]
@arabold did you identified a pattern triggering this crash ?

**Comment by arabold** [comment]
Not really, @calve. I just keep it running for a while. Usually I get an I/O error after maybe 1-2 hours, sometimes sooner. It might be a periodic network problem, it might be something else. However, my expectation would be that the S3 output plugin simply retry the upload - or at least not bring down the whole Logstash instance :)

Here's the setting I'm using:
```
  s3 {
    aws_credentials_file => "/etc/logstash/conf.d/aws_credentials.yml"
    bucket => "xxxxxx"
    prefix => "logstash/"
    region => "us-east-1"
    codec => "json_lines"
    size_file => 50000000
    time_file => 60
  }
```

**Comment by javifr** [comment]
I think that @trico and I have found the problem, but not have the time to write the solution yet.

I've been developing a [Dropbox plugin](https://github.com/javifr/logstash-plugin-output-dropbox) strongly based on the s3 plugin, and we are facing the same error.

I think that in some random cases, the **periodic_rotation_thread** :
```ruby
# this thread is doing its thing every certain minutes
 @periodic_rotation_thread = Stud::Task.new do
      LogStash::Util::set_thread_name("<Dropbox periodic uploader")

      Stud.interval(periodic_interval, :sleep_then_run => true) do
        @logger.debug("Dropbox: time_file triggered, bucketing the file", :filename => @tempfile.path)

        move_file_to_bucket_async(@tempfile.path)
        next_page
        create_temporary_file   # This function will close the temporary file
      end
    end
```

is closing a temp file with the create_temporary_file function:

```ruby
def create_temporary_file
    filename = File.join(@temporary_directory, get_temporary_filename(@page_counter))

    @logger.debug("Dropbox: Creating a new temporary file", :filename => filename)

    @file_rotation_lock.synchronize do
      unless @tempfile.nil?
        @tempfile.close
      end

      @tempfile = File.open(filename, "a")
    end
  end
```

that the main thread:
```ruby
 def handle_event(encoded_event)
    if write_events_to_multiple_files?
      if rotate_events_log?
     ...
```

with the **rotate_events_log?** function is trying to measure.
```ruby
  def rotate_events_log?
    @tempfile.size > @size_file
  end
```

So at the same given moment the time interval thread is closing a file that the handle_event function of the main thread is trying to measure.

```ruby
def create_temporary_file
    filename = File.join(@temporary_directory, get_temporary_filename(@page_counter))

    @logger.debug("Dropbox: Creating a new temporary file", :filename => filename)

    @file_rotation_lock.synchronize do
      unless @tempfile.nil?
        @tempfile.close
      end

      # HERE IS HAPPENING THE IO: CLOSED STREAM ERROR
      # Because the main thread is trying to issue a @tempfile.size of a file that is closed
      # And also because the shared var accross threads called @tempfile has not change its value yet

      @tempfile = File.open(filename, "a")
    end
  end
```

I guess the solution is to use some mutex system to control if the file is open or not in order to allow certain things.

1-New mutex:
```ruby
@file_measure_lock = Mutex.new
```
2-Protect the handle_event 
```ruby
def handle_event(encoded_event)
  @file_measure_lock.synchronize do
    if write_events_to_multiple_files?
...
```
3-Protect the create_temporary_file in the time thread:
```ruby
def configure_periodic_rotation
    @periodic_rotation_thread = Stud::Task.new do
      LogStash::Util::set_thread_name("<Dropbox periodic uploader")

      Stud.interval(periodic_interval, :sleep_then_run => true) do
        @logger.debug("Dropbox: time_file triggered, bucketing the file", :filename => @tempfile.path)

        move_file_to_bucket_async(@tempfile.path)
        next_page
       @file_measure_lock.synchronize do    # HERE
           create_temporary_file   
```

What do you think about it?

**Comment by ph** [comment]
@javifr Yes this is probably the cause and it make a lot of sense to me.

But we don't need to create a new mutex we can just reuse the `file_rotation_lock` inside the `#rotate_events_log` method to protect the `.size` check, we already use it to protect the writing and the file rotation? What do you think?

**Comment by javifr** [comment]
So do you think that with something like this we would be fine?

```ruby
def rotate_events_log?
    @file_rotation_lock.synchronize do
       @tempfile.size > @size_file
    end
end
```

Sounds good! I'll try to test it next week!

**Comment by ph** [comment]
@javifr yes, this is what I mean! thanks for testing this out.

**Comment by Onalu** [comment]
@javifr is there any update on this issue? I have this problem too.

**Comment by javifr** [comment]
Hi @Onalu ,I tested it on the plugin I am developing and for the moments the issue is solved, check it out:
https://github.com/javifr/logstash-plugin-output-dropbox/blob/master/lib/logstash/outputs/dropbox.rb#L251

**Comment by ph** [comment]
@Onalu I'll make a PR with that fix and update this thread.

**Comment by Onalu** [comment]
Thank you @javifr and @ph.

**Comment by guyboertje** [comment]
Closed by #20
